fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreTest.java

```diff
@@ -3648,56 +3651,140 @@ public void testIndexStateFormatVersionUpgrade() {
+            recordStore.markIndexDisabled(index.getName());
+            verifyOldAndNewIndexState(context, index, IndexState.DISABLED, FDBRecordStore.SAVE_INDEX_STATE_WITH_SUBSPACE_KEY_FORMAT_VERSION - 1);
+            commit(context);
+        }
+
+        // Upgrade with the renamed index. This must fail.
+        try (FDBRecordContext context = openContext()) {
+            final RecordMetaDataBuilder renamedBuilder = RecordMetaData.newBuilder().setRecords(TestNoIndexesProto.getDescriptor());
+            Index renamedIndex = new Index("index_renamed", "rec_no");
+            renamedIndex.setSubspaceKey(1);
+            renamedBuilder.addIndex("MySimpleRecord", renamedIndex);
+            FDBRecordStore.Builder storeBuilderRenamed = FDBRecordStore.newBuilder().setMetaDataProvider(renamedBuilder).setKeySpacePath(path);
+            renamedBuilder.setVersion(builder.getVersion());
+            MetaDataException exception = assertThrows(MetaDataException.class, () ->
+                    recordStore = storeBuilderRenamed.setContext(context)
+                            .setFormatVersion(FDBRecordStore.MAX_SUPPORTED_FORMAT_VERSION)
+                            .open());
+            assertEquals(exception.getMessage(), "Index index not defined");
+            commit(context);
+        }
+
+        // Now upgrade. The index state must upgrade.
+        try (FDBRecordContext context = openContext()) {
+            recordStore = storeBuilder.setContext(context)
+                    .setFormatVersion(FDBRecordStore.MAX_SUPPORTED_FORMAT_VERSION)
+                    .open();
+            verifyOldAndNewIndexState(context, index, IndexState.DISABLED, FDBRecordStore.MAX_SUPPORTED_FORMAT_VERSION);
+            commit(context);
+        }
+
+        // Now upgrade with the renamed index. This must fail.
+        try (FDBRecordContext context = openContext()) {
+            final RecordMetaDataBuilder renamedBuilder = RecordMetaData.newBuilder().setRecords(TestNoIndexesProto.getDescriptor());
+            Index renamedIndex = new Index("index_renamed", "rec_no");
+            renamedIndex.setSubspaceKey(1);
+            renamedBuilder.addIndex("MySimpleRecord", renamedIndex);
+            renamedBuilder.setVersion(builder.getVersion());
+            FDBRecordStore.Builder renamedStoreBuilder = FDBRecordStore.newBuilder().setMetaDataProvider(renamedBuilder).setKeySpacePath(path);
+            recordStore = renamedStoreBuilder.setContext(context)
+                    .setFormatVersion(FDBRecordStore.MAX_SUPPORTED_FORMAT_VERSION)
+                    .open();
+            // TODO: This is a bug. Querying the state of the index with the old and new names returns different states due to caching.
+            assertEquals(IndexState.DISABLED, recordStore.getRecordStoreState().getState(renamedIndex.getName())); // renamed index
+            assertEquals(IndexState.READABLE, recordStore.getRecordStoreState().getState(index.getName())); // old index state
```

Comment:
Hm, wait, isn't this correct? The renamed index now is marked as `DISABLED`, which is correct as the index at that subspace key is (still) disabled. The old index name...is now `READABLE`, whereas arguably it should be "index not found", but that's not necessarily "wrong" if the contract is "indexes outside the meta-data are always readable", which was I believe the contract before.

I think the bug I was thinking of requires using the `MetaDataVersionStampStoreStateCache` *and* for there to be an index name. In that case, I would expect `renamedIndex.getName()` to return `READABLE` and `index.getName()` to return `DISABLED` (i.e., the reverse of these results), and that would be incorrect.

Reply:
> The old index name...is now READABLE, whereas arguably it should be "index not found",

Good point. This is actually `not found`, and that's why it's returning `READABLE`. Given the planned changes for index meta-data, it's better to keep the contract as it was before, and decide on the eventual contract there. But you are right, this won't exercise the test case you had in mind. I've never used `MetaDataVersionStampStoreStateCache` before, let me look into it.

Reply:
Ah, yes, the new test looks better.